### ISSUE TITLE
Dinov2 Feature Extraction Code

### ISSUE THREAD

**Issue by michaelperez023** [pangzhan27/CMeRT#4]
Thank you for your interesting paper and releasing the code. I would appreciate any help you can give with this question:

I recreated your ViT-g/14 features for THUMOS’14 using DINOv2 from torch.hub and 24fps len6_pos4_stride6, and I’m getting cosine similarity ~0.989 to your published features. Could you share the exact feature extraction script you used or help me fix my feature extraction code?

for the 1st frame of video_test_0000004.mp4 of THUMOS14 with 224x224 input resolution, in the google drive folder (https://drive.google.com/drive/folders/1iAMCydhppFP_YgNIxKxDyMGd_Sarvy8h?usp=drive_link) your feature is:
[ 2.2567272 ,  0.39954942, -1.9204221 , ..., -1.3304411 , -1.244792  ,  1.5494777 ]

Mine is:
[ 2.4718065 ,  0.5115981 , -1.8707974 , ..., -1.4518318 , -1.1230706 ,  1.6531316 ]

My feature extraction code:
```import argparse
import cv2
import numpy as np
import torch
from torchvision import transforms
from PIL import Image

parser = argparse.ArgumentParser()
parser.add_argument("--video", required=True)
parser.add_argument("--img_size", type=int, default=224, choices=[224, 448])
parser.add_argument("--device", type=str, default="cuda")
args = parser.parse_args()

device = torch.device(args.device if torch.cuda.is_available() else "cpu")
model = torch.hub.load("facebookresearch/dinov2", "dinov2_vitg14").eval().to(device)

cap = cv2.VideoCapture(args.video)
if not cap.isOpened():
    raise RuntimeError("Could not open video")
num_frames = int(cap.get(cv2.CAP_PROP_FRAME_COUNT))
fps = float(cap.get(cv2.CAP_PROP_FPS))
print(f"Video {args.video}: {num_frames} frames @ {fps:.3f} fps")

duration = num_frames / fps
t = np.arange(int(np.floor(duration * 24.0 + 1e-6))) / 24.0
idx_24 = np.clip((t * fps).round().astype(int), 0, num_frames - 1)

centers = []
for s in range(0, len(idx_24) - 6 + 1, 6):
    centers.append(int(idx_24[s + 3]))   # pos=4 → idx=3
print("Using", len(centers), "frames after fps+chunking")

resize_size = 256 if args.img_size == 224 else 518
transform = transforms.Compose([
    transforms.Resize(resize_size, interpolation=transforms.InterpolationMode.BICUBIC),
    transforms.CenterCrop(args.img_size),
    transforms.ToTensor(),
    transforms.Normalize(mean=(0.485,0.456,0.406), std=(0.229,0.224,0.225)),
])

feats = []
with torch.no_grad():
    for i in range(0, len(centers), 32):
        batch_idx = centers[i:i+32]
        imgs = []
        for idx in batch_idx:
            cap.set(cv2.CAP_PROP_POS_FRAMES, int(idx))
            ok, frame = cap.read()
            if not ok: raise RuntimeError(f"Could not read frame {idx}")
            frame = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
            imgs.append(transform(Image.fromarray(frame)))
        batch = torch.stack(imgs).to(device)
        feats.append(model(batch).cpu().numpy())

feats = np.concatenate(feats, 0)
print(feats[0])

breakpoint()

**Comment by pangzhan27** [comment]
Sorry for the late reply, and thanks for your interest. 
I’ve roughly checked your code and have a few suggestions:
1. Make sure all your videos are at 24 FPS (you’ll need to preprocess them). Otherwise, the chunk size of 6 won’t make sense.
2. Ensure the resize dimensions match what we use, either 224 or 448.
3. For the frames in a chunk indexed 1–6, we use the 4th frame for feature extraction, which I believe you are currently doing correctly.
I’ve uploaded the related code to the drive for your reference. Hope it helps!
https://drive.google.com/drive/folders/1_UrckihiDDE0dgBWYHI6hldXh2AvKeOh?usp=sharing

**Comment by michaelperez023** [comment]
Thank you very much for the detailed reply and for sharing your feature extraction code. Using your code I was able to extract features with 0.999 cosine similarity to the extracted features you released. I attribute the remaining difference to different environments (machine/CUDA/PyTorch versions). I was also able to correct my own feature extraction code to match yours.

I really appreciate your help, this resolved my issue completely.
